File: day10_monitoring_station.cpp
// Day 10. Monitoring Station
// https://adventofcode.com/2019/day/10
#include<iostream>
#include<fstream>
#include<set>
#include<cmath>

using namespace std;

char EMPTY = '.';
char ASTEROID = '#';
double INF = (double) INT_MAX;

auto read_input() {
    string line;
    vector<vector<char>> grid;
    ifstream input("inputs/day10_input.txt");
    while (getline(input, line)) {
        vector<char> l(line.begin(), line.end());
        grid.push_back(l);
    }
    input.close();
    return grid;
}


// Map of all asteroids
// best location is the asteroid that can detect the largest number of other asteroids
// Build a relationship map. If A can see B, then B can see A
// Exact ray tracing out of each asteroid
// Create a map of slopes for each asteroid vs all other asteroids
// Get min distance on each slope

struct Coord {
    int x;
    int y;
};


void printCoord(Coord& c) {
    cout << "Coord: (" << c.x << ", " << c.y << ")" << endl;
}


double calcSlope(Coord& c1, Coord& c2) {
    if (c1.x - c2.x == 0) { return INF; }
    return (double) (c1.y - c2.y) / (double) (c1.x - c2.x);
}


int canSee(Coord& self, vector<Coord>& asteroids) {
    // count number of unique slopes, but one in each direction
    
    // keep track of slopes for asteroids above self and below self separately
    set<double> above;
    set<double> below;
    set<double> left;
    set<double> right;
    double slope;

    for (auto a : asteroids) {
        slope = calcSlope(self, a);
        if (a.y == self.y && a.x == self.x) {
            continue; // self
        } else if (a.y > self.y) {
            above.insert(slope);
        } else if (a.y < self.y) {
            below.insert(slope);
        } else if (a.x < self.x) {
            // same line y's equal
            right.insert(slope);
        } else {
            left.insert(slope);
        }
    }

    return above.size() + below.size() + left.size() + right.size();
}


void part1() {
    auto grid = read_input();
    vector<Coord> asteroids;

    for (int y=0; y<grid.size(); y++) {
        for (int x=0; x<grid[y].size(); x++) {
            if (grid[y][x] == ASTEROID) {
                Coord c{x, y};
                asteroids.push_back(c);
            }
        }
    }

    Coord mxcoord;
    int mxcount = 0;
    int count;
    for (auto self : asteroids) {
        count = canSee(self, asteroids);
        if (count > mxcount) {
            mxcount = count;
            mxcoord = self;
        }
        
    }

    cout << "Solution 1: " << mxcount << endl;
    cout << "Best Location: (" << mxcoord.x << ", " << mxcoord.y << ')' << endl;

}



int calcMult(Coord& c) {
    cout << "Calculate: " << c.x << ' ' << c.y << endl;
    return 100 * c.x + c.y;
}


int lazer(Coord station, vector<Coord>& asteroids) {
    // keep track of slopes for asteroids above self and below self separately
    unordered_map<double, vector<Coord>> above;
    unordered_map<double, vector<Coord>> below;
    unordered_map<double, vector<Coord>> right;
    unordered_map<double, vector<Coord>> left;
    double slope;

    for (auto a : asteroids) {
        slope = calcSlope(station, a);
        if (a.y == station.y && a.x == station.x) {
            continue; // station
        }
        
        if (a.x > station.x) {
            right[slope].push_back(a);
        } else if (a.x < station.x) {
            left[slope].push_back(a);
        } else if (a.y < station.y) {
            // same line y's equal
            above[slope].push_back(a);
        } else {
            below[slope].push_back(a);
        }
    }

    vector<double> rightslopes;
    vector<double> leftslopes;


    int x0 = station.x;
    int y0 = station.y;

    for (auto p : right) {
        rightslopes.push_back(p.first);
        // sort asteroids by distance from station
        sort(p.second.begin(), p.second.end(),
            [](const Coord& c1, const Coord& c2) {
                return (abs(c1.x - 11) + abs(c1.y - 13)) > (abs(c2.x-11) + abs(c2.y-13));
            });
    }
    sort(rightslopes.begin(), rightslopes.end(), greater<>());
        

    for (auto p: left) {
        leftslopes.push_back(p.first);
        sort(p.second.begin(), p.second.end(),
            [](const Coord& c1, const Coord& c2) {
                return (abs(c1.x - 11) + abs(c1.y - 13)) > (abs(c2.x-11) + abs(c2.y-13));
            });
    }
    sort(leftslopes.begin(), leftslopes.end(), greater<>());

    // Sort each direction and slope in order

    // Lazer
    int count = 0;
    Coord curr;

    // Rotate clockwise
    while (count < 200) {
        cout << "Count: " << count << endl;
        // 12 oclock
        if (above[INF].size() > 0) {
            printCoord(above[INF].back());
            if (++count == 200) {
                return calcMult(above[INF].back());
            }
            above[INF].pop_back();
        }

        for (auto s : rightslopes) {
            if (right[s].size() > 0) {
                printCoord(right[s].back());
                if (++count == 200) {
                    return calcMult(right[s].back());
                }
                right[s].pop_back();
            }
        }

        // 6 oclock
        if (below[INF].size() > 0) {
            printCoord(below[INF].back());
            if (++count == 200) {
                return calcMult(below[INF].back());
            }
            below[INF].pop_back();
        }

        for (auto s : leftslopes) {
            if (left[s].size() > 0) {
                printCoord(left[s].back());
                if (++count == 200) {
                    return calcMult(left[s].back());
                }
                left[s].pop_back();
            }
        }
    }

    return -1;
}




void part2() {
    auto grid = read_input();
    vector<Coord> asteroids;

    for (int y=0; y<grid.size(); y++) {
        for (int x=0; x<grid[y].size(); x++) {
            if (grid[y][x] == ASTEROID) {
                Coord c{x, y};
                asteroids.push_back(c);
            }
        }
    }

    // from Part 1
    Coord station{11, 13};

    // Laser starts at 12 o'clock and moves clockwise
    //
    int res = lazer(station, asteroids);
    cout << "Solution 2: " << res << endl;

}


int main() {
    part1();
    part2();
}

File: day10_monitoring_station.py
# Day 10. Monitoring Station
# https://adventofcode.com/2019/day/10D

from collections import defaultdict

def read_input():
    grid = []
    with open('inputs/day10_input.txt') as f:
        lines = f.readlines()
        for l in lines:
            grid.append(list(l.replace('\n', '')))
        
    return grid


ASTEROID = '#'
SPACE = '.'

INF = float('inf')

def calc_slope(r1, c1, r2, c2):
    if (c2-c1) == 0: return INF
    return -1 * (r2-r1) / (c2-c1)

def calc_dist(x1, y1, x2, y2):
    return abs(x1-x2) + abs(y1-y2)


def get_asteroids(grid):
    asteroids = []
    for r in range(len(grid)):
        for c in range(len(grid[0])):
            if grid[r][c] == ASTEROID:
                asteroids.append((r, c))
                
    return asteroids


def buildSeen(r, c, asteroids):
    right = defaultdict(list)
    left = defaultdict(list)
    above = defaultdict(list)
    below = defaultdict(list)
    
    for r1, c1 in asteroids:
        slope = calc_slope(r, c, r1, c1)
        if (r == r1 and c == c1):
            continue
        elif (c == c1 and r > r1):
            above[slope].append((r1, c1))
        elif (c == c1 and r < r1):
            below[slope].append((r1, c1))
        elif (c1 > c):
            right[slope].append((r1, c1))
        elif (c1 < c):
            left[slope].append((r1, c1))
        else:
            raise ValueError(f"Unknown case - {r} - {c} - {r1} - {c1}")


    return right, left, above, below
    





def part1():
    grid = read_input()
    asteroids = get_asteroids(grid)

    mxcount = 0
    mxr, mxc = 0, 0
    for r, c in asteroids:
        right, left, above, below = buildSeen(r, c, asteroids)

        count = len(right.keys()) + len(left.keys()) + len(above.keys()) + len(below.keys())
        if count > mxcount:
            # print(r, c, count)
            mxcount = count
            mxr = r
            mxc = c
            
    return mxcount, mxr, mxc

    



def part2():
    grid = read_input()
    asteroids = get_asteroids(grid)
    _, r, c = part1()
    right, left, above, below = buildSeen(r, c, asteroids)

    rightkeys = sorted(right.keys(), reverse=True)
    leftkeys = sorted(left.keys(), reverse=True)
    
    # Sort by distance from 
    for d in (right, left, above, below):
        for k, v in d.items():
            v.sort(key=lambda x: calc_dist(x[0], x[1], r, c) * -1)
            
            
            
    order = []
    while len(order) < 200:
        # 12 oclock
        if above[INF]:
            v = above[INF].pop()
            # print(INF, v)
            order.append(v)
        
        
        # right
        for k in rightkeys:
            if right[k]:
                v = right[k].pop()
                # print(k, v)
                order.append(v)

        # 6 oclock
        if below[INF]:
            v = below[INF].pop()
            # print(INF, v)
            order.append(v)
            
        # left
        for k in leftkeys:
            if left[k]:
                v = left[k].pop()
                # print(k, v)
                order.append(v)
            
        
    y, x = order[199]
    # print(order)
    # print(order[199])
    return x * 100 + y 
        
    

    

                
    




if __name__ == '__main__':
    count, _, _ = part1()
    print("Part 1", count)
    res = part2()
    print("Part 2", res)